### PR TITLE
Responsive mode JSUI-2269

### PR DESCRIPTION
**Pull request by louis-bompart**
This PR goal is to allow to create a Search Interface and enforce its layout to the one of a large, medium or small screen without doing some trickery like setting the responsive breakpoints to absurd values.

I chose to do the change like this (i.e. return `POSITIVE_INFINITY` and 0 depending on the responsive mode) to avoid massive changes in the responsive components.

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/SearchInterface/SearchInterface.ts [hunk 1]
```diff
@@ -439,7 +440,17 @@ export class SearchInterface extends RootComponent implements IComponentBindings
...
-    })
+    }),
+    /**
+     * Specifies the search interface responsive mode that should be used.
+     *
+     * When the mode is auto, the width of the window/device that displays the search page is used to determine which layout the search page should use (see [enableAutomaticResponsiveMode]{@link SearchInterface.options.enableAutomaticResponsiveMode}, [responsiveMediumBreakpoint]{@link SearchInterface.options.responsiveMediumBreakpoint} and [responsiveSmallBreakpoint{@link SearchInterface.options.responsiveSmallBreakpoint}])
+     *
+     * When it's not on auto, the width is ignored and the the layout used depends on this option (e.g. If set to "small", then the search interface layout will be the same as if it was on a narrow window/device)
+     */
+    responsiveMode: ResponsiveModes[(ComponentOptions.buildStringOption({
```
Comment (olamothe) [inline src/ui/SearchInterface/SearchInterface.ts]: I'd change that to : 
```
responsiveMode: ComponentOptions.buildCustomOption<ValidResponsiveMode>((value)=> {
  // Validator function for the string passed, verify it's one of the accepted values.
  // If valid :
  return value
// else : 
warningMessage();
return 'auto';
}, {
 defaultValue : 'auto'
})
```

**Comment by louis-bompart** [comment]
Changes done, UT added. Travis will stay red due to Veracode  ¯\_(ツ)_/¯

File: src/ui/SearchInterface/SearchInterface.ts [hunk 2]
```diff
@@ -440,7 +446,28 @@ export class SearchInterface extends RootComponent implements IComponentBindings
     responsiveSmallBreakpoint: ComponentOptions.buildNumberOption({
       defaultValue: SMALL_SCREEN_WIDTH,
       depend: 'enableAutomaticResponsiveMode'
-    })
+    }),
+    /**
+     * Specifies the search interface responsive mode that should be used.
+     *
+     * When the mode is auto, the width of the window/device that displays the search page is used to determine which layout the search page should use (see [enableAutomaticResponsiveMode]{@link SearchInterface.options.enableAutomaticResponsiveMode}, [responsiveMediumBreakpoint]{@link SearchInterface.options.responsiveMediumBreakpoint} and [responsiveSmallBreakpoint{@link SearchInterface.options.responsiveSmallBreakpoint}])
```
Comment (ThibodeauJF) [inline src/ui/SearchInterface/SearchInterface.ts]: That's a pretty long line, could you put the (see [...]) part on a new line?

File: src/ui/SearchInterface/SearchInterface.ts [hunk 3]
```diff
@@ -440,7 +446,28 @@ export class SearchInterface extends RootComponent implements IComponentBindings
...
+     *
+     * When it's not on auto, the width is ignored and the the layout used depends on this option (e.g. If set to "small", then the search interface layout will be the same as if it was on a narrow window/device)
```
Comment (ThibodeauJF) [inline src/ui/SearchInterface/SearchInterface.ts]: Same here, newline for (e.g. [...]) it makes reviews on github a little more friendly. Thanks!

File: src/ui/SearchInterface/SearchInterface.ts (hunk 2 above)
Comment (louis-bompart) [reply src/ui/SearchInterface/SearchInterface.ts]: Done 😄

File: src/ui/SearchInterface/SearchInterface.ts [hunk 4]
```diff
@@ -440,7 +446,28 @@ export class SearchInterface extends RootComponent implements IComponentBindings
...
+     */
+    responsiveMode: ComponentOptions.buildCustomOption<ValidResponsiveMode>(
+      value => {
+        // Validator function for the string passed, verify it's one of the accepted values.
+        if (value === 'auto' || value === 'small' || value === 'medium' || value === 'large') {
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply src/ui/SearchInterface/SearchInterface.ts]: That was an `enum` but have been changed to a type following @olamothe request.

File: unitTests/ui/ResponsiveComponentsTest.ts [hunk 5]
```diff
@@ -79,10 +101,81 @@ export function ResponsiveComponentsTest() {
...
+      });
+
+      it('should return isLargeScreenWidth', () => {
+        expect(responsiveComponent.isLargeScreenWidth()).toBe(true);
+      });
+    });
+
+    describe('with a small responsiveMode and auto responsive mode', () => {
+      beforeEach(() => {
+        windoh['clientWidth'] = 600;
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply unitTests/ui/ResponsiveComponentsTest.ts]: This `describe` is bad, I'll change it.
I'll also replace those number by thing like `SMALL_SCREEN_WIDTH+1` it should be more intuitive.

File: unitTests/ui/ResponsiveComponentsTest.ts [hunk 6]
```diff
@@ -79,10 +101,81 @@ export function ResponsiveComponentsTest() {
...
+        responsiveComponent = new ResponsiveComponents(windoh);
+        responsiveComponent.setResponsiveMode('small');
+      });
+
+      it('should return 0 for medium breakpoint and positive infinity for small', () => {
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply unitTests/ui/ResponsiveComponentsTest.ts]: Will be split.

File: unitTests/ui/ResponsiveComponentsTest.ts [hunk 7]
```diff
@@ -79,10 +101,81 @@ export function ResponsiveComponentsTest() {
...
+        expect(responsiveComponent.getSmallScreenWidth()).toBe(Number.POSITIVE_INFINITY);
+        expect(responsiveComponent.getMediumScreenWidth()).toBe(0);
+      });
+
+      it('should return isSmallScreenWidth', () => {
+        expect(responsiveComponent.isSmallScreenWidth()).toBe(true);
+      });
+
+      it('should not return isMediumScreenWidth', () => {
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply unitTests/ui/ResponsiveComponentsTest.ts]: Changed

**Pull request merged**